src/appinsights.ts

```diff
@@ -166,3 +166,48 @@ export function initAppInsights(
     ...agentOpts
   });
 }
+
+// tslint:disable-next-line: no-any
+const noop = (..._: readonly any[]): void => void 0;
+export interface ITelemetryDefaultClient {
+  addTelemetryProcessor: typeof appInsights.defaultClient.addTelemetryProcessor;
+  clearTelemetryProcessors: typeof appInsights.defaultClient.addTelemetryProcessor;
+  flush: typeof appInsights.defaultClient.flush;
+  track: typeof appInsights.defaultClient.track;
+  trackAvailability: typeof appInsights.defaultClient.trackAvailability;
+  trackDependency: typeof appInsights.defaultClient.trackDependency;
+  trackEvent: typeof appInsights.defaultClient.trackEvent;
+  trackException: typeof appInsights.defaultClient.trackException;
+  trackMetric: typeof appInsights.defaultClient.trackMetric;
+  trackNodeHttpDependency: typeof appInsights.defaultClient.trackNodeHttpDependency;
+  trackNodeHttpRequest: typeof appInsights.defaultClient.trackNodeHttpRequest;
+  trackNodeHttpRequestSync: typeof appInsights.defaultClient.trackNodeHttpRequestSync;
+  trackPageView: typeof appInsights.defaultClient.trackPageView;
+  trackRequest: typeof appInsights.defaultClient.trackRequest;
+  trackTrace: typeof appInsights.defaultClient.trackTrace;
+}
+
+/**
+ * Wraps ApplicationInsights' defaultClient to provide a version which is never undefined.
+ * ApplicationInsights' defaultClient is instantiated as a singleton inside its module. If no env configuration is provided, this results to be undefined and forces user to check everytime before using it
+ * In case is not defined, we provide a dummy clone in which its methods results in no-operations, avoiding unhandled exceptions.
+ */
+export const defaultClient: ITelemetryDefaultClient = appInsights.defaultClient
```

Comment:
what about

```ts
export const defaultClient = new Proxy(appInsights.defaultClient || {}, {
  // tslint:disable-next-line: no-any
  get(target, prop): () => any {
    return prop in target
      ? // tslint:disable-next-line: no-any
        (target as any)[prop]
      : constVoid;
  }
});
```

Reply:
I've tried this for hours. The point is: in the case `appInsights.defaultClient` is undefined, you cannot instantiate a proxy for it.

Reply:
that's the reason of `|| {}`

Reply:
ok, I just realised I could have just done this:
```typescript
export const defaultClient: typeof appInsights.defaultClient = appInsights.defaultClient
  ? appInsights.defaultClient
  : (new Proxy(
      {},
      {
        // tslint:disable-next-line: no-any
        get(): () => void {
          return () => void 0;
        }
      }
    ) as typeof appInsights.defaultClient);
```

Reply:
another option is to have a method for each prop ie. that takes the client as input

```ts
trackEvent(defaultClient, opts: typeof Parameters<trackEvent) => 
     fromNullable(defaultClient).map(_ => tryCatch2v(() => _.trackEvent(opts), constVoid));
```

which catch the exception as well

Reply:
which is a bit incomplete, as `appInsights.defaultClient` also has some non-function props

Reply:
what's the problem with non function props?

Reply:
```ts
export const defaultClient =
  appInsights.defaultClient ??
  ((constVoid as unknown) as appInsights.TelemetryClient);
```
 ?

Reply:
they need to be stubbed with something different than a function

Reply:
> another option is to have a method for each prop ie. that takes the client as input
> 
> ```ts
> const trackEvent(defaultClient, opts: typeof Parameters<trackEvent) => 
>      fromNullable(defaultClient).map(_ => tryCatch2v(() => _.trackEvent(opts), constVoid));
> ```
> 
> which catch the exception as well. since we're wrapping the client it's probaly worth it

Do you think we should do it? I tried not to rewrite each method but I agree it might be useful

Reply:
yes, basically we use just a few (trackEvent, trackException), others may be added when needed.

Reply:
a nice way to avoid passing the client is this:

https://github.com/pagopa/io-functions-bonusapi/blob/0b6cbd493328396f6ec9e2489df833ff384963a4/utils/appinsights.ts#L30